### ISSUE TITLE
get_ai_assistance_feature severely degrades performance with uncached HTTP calls

### ISSUE THREAD

**Issue by tomjn** [Automattic/jetpack#33248]
### Impacted plugin

Jetpack

### Quick summary

When using the block editor, Jetpack makes 3 HTTP requests via `get_ai_assistance_feature` in the Jetpack AI helper class to fetch some data. These 3 HTTP requests are identical and take ~0.4-0.6s each, adding up to 1.5s slowdown on block editor screens during asset enquing.

Note that at no point had I attempted to use Jetpack AI features.

1. This should have been cached so only 1 HTTP request was needed
2. It shouldn't be necessary to do this for the Jetpack AI feature to work, nothing prevents this being done in a REST API endpoint so that it can be done at runtime in JS

### Steps to reproduce

Load a block editor screen with Jetpack activated.

### A clear and concise description of what you expected to happen.

I did not expect 3 slow uncached HTTP requests to happen in PHP to the WordPress.com API

### What actually happened

3 duplicate uncached HTTP requests were made for an AI feature I had not yet used.

### Impact

Most (> 50%)

### Available workarounds?

No but the platform is still usable

### Platform (Simple and/or Atomic)

Self-hosted

**Comment by tomjn** [comment]
This is the code responsible:

https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/lib/class-jetpack-ai-helper.php#L367-L392

I have an object cache and memcached, so this could have been stored in `WP_Cache` keyed against the IP, it would have been so fast I would not have noticed it and Jetpack would have gotten away with it.

Also why is it bundling the visitors IP? That seems redundant and concerning, it should at least include a comment explaining why if there's a good reason.

cc @retrofox who wrote the code in https://github.com/Automattic/jetpack/pull/30926

Noting that my suggestions for it using a REST API endpoint has actually already been half implemented as that PR added a `/ai-assistant-feature` endpoint, so I'm not sure why it's needed just so that some initial state can be set:

https://github.com/Automattic/jetpack/blob/add48c8bd56973fdd6d738ab2059cdf860269b54/projects/plugins/jetpack/class.jetpack-gutenberg.php#L680-L689

I can only see it used here:

https://github.com/Automattic/jetpack/blob/add48c8bd56973fdd6d738ab2059cdf860269b54/projects/plugins/jetpack/extensions/blocks/ai-assistant/edit.js#L47

Which would be an incorrect usage, it should be using the custom hook instead `useAIFeature`:

https://github.com/Automattic/jetpack/blob/add48c8bd56973fdd6d738ab2059cdf860269b54/projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-ai-feature/index.ts#L25

That hook would be an ideal place to check if the data has been fetched and query the REST API which appears to be exactly what it does:

https://github.com/Automattic/jetpack/blob/add48c8bd56973fdd6d738ab2059cdf860269b54/projects/plugins/jetpack/extensions/blocks/ai-assistant/hooks/use-ai-feature/index.ts#L38-L64

---

So if it's already trying to fetch it from the REST API, why do it in enqueue block assets multiple times?

**Comment by jeherve** [comment]
cc @Automattic/jetpack-agora Is this something you could look into?

Thank you!

**Comment by tomjn** [comment]
I think a part of this too was the assumption that the enqueue block assets hook is called only once, though even than adding 0.4/0.5s to the load time is problematic ( though not as bad as ~1.5s)

**Comment by retrofox** [comment]
Let's find a way to improve the performance when request AI feature data

**Comment by retrofox** [comment]
I've created a PR that avoids performing the AI assistant feature request, but I don't know if that's okay. The idea behind requesting the data when the AI Assistant block mounts, for instance, is that we'd like to know the number of requests at that moment. This is something that will be useful in the short/medium term, I guess.

When we started to implement the AI Assistant block, we wanted to do the following UI implementation:

<img width="301" alt="image" src="https://github.com/Automattic/jetpack/assets/77539/93bb65b5-2061-4566-88f6-3f82b38a0005">




> I think a part of this too was the assumption that the enqueue block assets hook is called only once, though even than adding 0.4/0.5s to the load time is problematic ( though not as bad as ~1.5s)

If I'm not wrong, there is only one sync request that would add the slowdown. The rest of the requests are async, right?

**Comment by tomjn** [comment]
> If I'm not wrong, there is only one sync request that would add the slowdown. The rest of the requests are async, right?

None were async, all three were blocking requests, but even then only one request should have been necessary. The fundamental problem there is that it's assumed the hook only fires once, which is sometimes the case, therefore it will only make the request once, but that's a broken assumption. I understand you want fresh data but the likelihood the data will change on the timescale of 0.1-0.2s is highly improbable, ( and if you did need to make 3 calls, would it not make more sense to eliminate the first two and only use and make the call that happens last? )

Even if it did only fire once though, it's still half a second of blocking time trying to generate initial data store values.

Note that when I mention caching, I don't mean persistent caching, it would be a big improvement to assign the result to a variable or static variable that only survives for that request. That way you could call the function as many times as you want and be guaranteed only 1 request per request.

> When we started to implement the AI Assistant block, we wanted to do the following UI implementation:

For this I think a PHP based request would be incorrect, in addition to slower. If I use a Jetpack AI feature then the data is immediately invalidated and has to be requested again, defeating the point of fetching it in PHP. Additionally the data is unnecessary until I open the AI Assistant panel, so it would be more efficient both in the browser, the server, and at Automattics servers, if it fetched the data async in JS via the REST API.

**Comment by tomjn** [comment]
As noted on https://github.com/Automattic/jetpack/pull/33319 there's been a big misunderstanding.

The problem is PHP making multiple blocking HTTP requests to an A8C server during block asset enqueing. Refreshing data in React was actually my suggestion of how to fix this, not the problem.

**Comment by retrofox** [comment]
Got it.
Created a new PR https://github.com/Automattic/jetpack/pull/33391 with a different approach.
It avoids requesting the dotcom endpoint for 10 seconds by caching AI Assistant feature data.
